src/System.Composition.Convention/src/Microsoft/Internal/Lock.Reader.cs

```diff
@@ -0,0 +1,32 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading;
+
+namespace Microsoft.Internal
+{
+    internal struct ReadLock : IDisposable
+    {
+        private readonly Lock _lock;
+        private int _isDisposed;
+
+        public ReadLock(Lock @lock)
+        {
+            _isDisposed = 0;
+            _lock = @lock;
+            _lock.EnterReadLock();
+        }
+
+        public void Dispose()
+        {
+            if (Interlocked.CompareExchange(ref _isDisposed, 1, 0) == 0)
```

Comment:
Definitely could be a footgun in waiting. So long as usage is always as per:

```csharp
using (new ReadLock(_lock))
{
    _memberInfos.TryGetValue(memberInfo, out cachedAttributes);
}
```

...then it's benign, but switching this and `WriteLock` to `class`es would be a reasonable change as well.

This code when through some pretty heavy profiling/optimisation - the original goal was to get composition such as `new A(new B())` or `new A(B.Instance)` in line with the native/hand-coded equivalents. (Other IoC containers under development at the time were also heading for this goal.)

Garbage on the heap was a big barrier to to it - but without the harness/original numbers I can't say just how much. @dsplaisted are there any perf tests with this that might point to how much the switch to `class`es would cost?